fedimint-api/src/task.rs

```diff
@@ -1,23 +1,164 @@
 use std::future::Future;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering::SeqCst;
+use std::sync::Arc;
 use std::time::{Duration, Instant};
 
+use anyhow::anyhow;
+use futures::lock::Mutex;
+pub use imp::*;
 use thiserror::Error;
+#[cfg(not(target_family = "wasm"))]
+use tokio::task::JoinHandle;
+use tracing::info;
+#[cfg(target_family = "wasm")]
+type JoinHandle<T> = futures::future::Ready<anyhow::Result<T>>;
 
 #[derive(Debug, Error)]
 #[error("deadline has elapsed")]
 pub struct Elapsed;
 
+#[derive(Debug, Default)]
+struct TaskGroupInner {
+    /// Was the shutdown requested, either externally or due to any task failure?
+    is_shutting_down: AtomicBool,
+    join: Mutex<Vec<(String, JoinHandle<()>)>>,
+}
+
+/// A group of task working together
+///
+/// Using this struct it is possible to spawn one or more
+/// main thread collabarating, which can cooperatively gracefully
+/// shut down, either due to external request, or failure of
+/// one of them.
+///
+/// Each thread should periodically check [`TaskHandle`] or rely
+/// on condition like channel disconnection to detect when it is time
+/// to finish.
+#[derive(Clone, Default, Debug)]
+pub struct TaskGroup {
```

Comment:
Really cool abstraction. I wonder if we still need a way to get output from tasks in some places? Maybe not for now.

Reply:
It could be done, where `TaskHandle` could give out some `T: Write` to places. Though as we are doing logging via `tracing` it really wants to do all relevant output to a global `stderr` so it might be hard.